include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,251 @@
+    };
+
+    struct time_data_t {
+      uint64_t num_timings_ = 0;
+      timing_t value_ = 0;
+    };
+
+    struct tuner_t {
+      std::mutex m_;
+
+      std::chrono::high_resolution_clock::time_point t0_;
+
+      timing_t best_timing_ = std::numeric_limits<timing_t>::max();
+      resource_with_offset_t best_resource_;
+
+      const size_type max_resource_to_profile_;
+      uint64_t next_resource_to_profile_ = 0; // as offset in resources
+
+      using time_by_offset_t = std::map<size_type, time_data_t>;
+      time_by_offset_t time_by_offset_;
+
+      double resample_time_ = 0;
+
+      tuner_t(resource_with_offset_t br, size_type resources_size, double rt) 
+        : t0_(std::chrono::high_resolution_clock::now()), 
+          best_resource_(br), 
+          max_resource_to_profile_(resources_size),
+          resample_time_(rt) {}
+
+      size_type get_resource_to_profile() {
+        std::unique_lock<std::mutex> l(m_);
+        if (next_resource_to_profile_ < max_resource_to_profile_) {
+          return next_resource_to_profile_++;
+        } else if (resample_time_ == never_resample) {
+          return use_best_resource;
+        } else {
+          auto now = std::chrono::high_resolution_clock::now();
+          auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now-t0_).count();
+          if (ms < resample_time_) {
+            return use_best_resource;
+          } else {
+            t0_ = now;
+            return next_resource_to_profile_ = 0;
+          }
+        }
+      }
+
+      // called to add new profile info
+      void add_new_timing(resource_with_offset_t r, timing_t t) {
+        std::unique_lock<std::mutex> l(m_);
+        auto offset = r.offset_;
+        timing_t new_value = t;
+        if (time_by_offset_.count(offset) == 0) {
+          time_by_offset_[offset] = time_data_t{1, t};
+        } else {
+          auto &td = time_by_offset_[offset];
+          auto n = td.num_timings_;
+          new_value = (n*td.value_+t)/(n+1);
+          td.num_timings_ = n+1;
+          td.value_ = new_value;
+        }
+        if (new_value < best_timing_) {
+          best_timing_ = new_value;
+          best_resource_ = r;
+        }
+      } 
+
+    };
+
+   public:
+
+    // Needed by Policy Traits
+    using resource_type = decltype(unwrap(std::declval<wrapped_resource_t>()));
+    using wait_type = typename Backend::wait_type;
+    using selection_type = internal::auto_tune_selection_type<Backend, resource_with_offset_t, tuner_t, KeyArgs...>;
+
+    auto_tune_policy(double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(resample_time);
+      }
+    }
+
+    auto_tune_policy(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(u, resample_time);
+      }
+    }
+   
+    void initialize(double resample_time=never_resample) {
```

Comment:
Spec calls the argument to initialize() offset, so maybe we change the spec to something generic to account for different uses in different policies